### PR TITLE
[pytorch][cuda] Optimized softmax forward native CUDA implementation

### PR DESCRIPTION
**Pull request by valentinandrei**
In Triton's [softmax tutorial](https://triton-lang.org/main/getting-started/tutorials/02-fused-softmax.html), native performance is significantly lower than Triton's. We accelerated the native code as follows:

--> Wrote a CUDA kernel `cunn_SoftMaxForwardSmem` for softmax forward that caches the inputs in shared memory. Currently the maximum usable shared memory is 48KB to preserve compatibility with older generation Kepler GPUs but we can increase this. This kernel uses vectorized loads and stores and runs on problem sizes that fit in shared memory and use aligned buffers.

--> Modified the default implementation's intra thread block reduction to use warp shuffles as the first step in reduction and use shared memory only to reduce across warps.

--> Simplified the `WriteFpropResults` code because the loop unrolling brought no benefits but had a potentially detrimental effect on register usage.

We can observe that there is still an advantage in the Triton implementation. We were able to recover the gap by using native `__expf` but we decided to leave `std::exp` to avoid affecting numerical stability.

```
Tests are ran on an A100 GPU using the benchmark in the Triton tutorial.

Before

softmax-performance:
          N       Triton  Torch (native)  Torch (jit)
0     256.0   336.946021      595.781814   241.830261
1     384.0   737.741110      762.046555   297.890900
2     512.0   884.128199      860.899863   362.829080
3     640.0   936.228605      901.458039   376.211253
4     768.0  1005.024893      973.306952   384.187594
..      ...          ...             ...          ...
93  12160.0  1336.034308      858.096595   330.642735
94  12288.0  1339.248830      837.047196   331.146707
95  12416.0  1338.877891      839.317673   329.113513
96  12544.0  1335.383669      835.342136   330.067106
97  12672.0  1339.402120      821.690012   329.854051

After

softmax-performance:
          N       Triton  Torch (native)  Torch (jit)
0     256.0   375.833684      602.629893   237.019883
1     384.0   312.572329      739.127852   301.777431
2     512.0   495.546303      863.736375   368.438508
3     640.0   520.953881      884.426455   369.633391
4     768.0   677.374681      975.722054   385.317013
..      ...          ...             ...          ...
93  12160.0  1337.253933     1300.589124   330.655916
94  12288.0  1336.333052     1188.412588   331.116192
95  12416.0  1337.610105     1209.703474   329.232825
96  12544.0  1338.723893     1232.849225   330.003484
97  12672.0  1340.232227     1236.057117   329.925347
```

### CODE REVIEW

**Review by eqy — COMMENTED** [review]
Great! Do you have performance data for before/after this change? (I'm not sure if the torch native column in the original post refers to before or after this change)

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 1]
```diff
@@ -172,6 +173,19 @@ inline dim3 SoftMax_getBlockSize(int ILP, uint64_t dim_size) {
   return dim3(block_size);
 }
 
+inline dim3 SoftMaxForward_getBlockSize(uint64_t dim_size) {
+  uint64_t block_size = 1;
+  uint64_t max_block_size = std::min(dim_size / ILP, static_cast<uint64_t>(max_threads));
```
Comment (eqy) [inline aten/src/ATen/native/cuda/SoftMax.cu:178]: looks like the build is breaking because ILP isn't defined here (was it intended to be an argument)?

Comment (valentinandrei) [reply aten/src/ATen/native/cuda/SoftMax.cu]: Whops. Wrong copy paste from my dev machine. I'll spend some time this week to make the PR landable and make sure all the tests pass.

**Comment by valentinandrei** [comment]
> Great! Do you have performance data for before/after this change? (I'm not sure if the torch native column in the original post refers to before or after this change)

I added the `before` data. For the moment I'd take the `after` data with a grain of salt because I need to validate numerics (my tests pass but usually CI finds new stuff for these types of changes). Hopefully during the week we should be ok.

**Comment by Skylion007** [comment]
@valentinandrei You accidentally downgraded pybind11. Please fix.

**Comment by valentinandrei** [comment]
> @valentinandrei You accidentally downgraded pybind11. Please fix.

Thanks. Fixed.

**Comment by valentinandrei** [comment]
cc: @malfet @bertmaher

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 2]
```diff
@@ -440,6 +459,58 @@ ilpReduce(index_t shift,
...
+      threadMax = redFuncMax(threadMax, val);
+      threadExp = redFuncAdd(threadExp, std::exp(val));
+    }
+    size -= blockDim.x;
+    data += blockDim.x;
+  }
+  index_t last = size % (ILP * blockDim.x);
+
+  T v[ILP];
+  LoadT* value = reinterpret_cast<LoadT*>(&v);
```
Comment (malfet) [inline aten/src/ATen/native/cuda/SoftMax.cu:493]: Might not be a 100% correct code (but imo this is how what you are trying to do should be expressed)
But also, why do you need `LoadT` at all, is it because copying from vector to vector is faster?
```suggestion
  static_assert (sizeof(T)*ILP == sizeof(template LoadT));
  union {
     T v[ILP];
     LoadT value;
   } x;
```

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 3]
```diff
@@ -440,6 +459,57 @@ ilpReduce(index_t shift,
...
+  MaxFloat<T, AccumT> redFuncMax;
+  Add<AccumT> redFuncAdd;
+
+  // Prologue
+  index_t offset = threadIdx.x;
+  if (shift > 0) {
+    data -= shift;
+    size += shift;
+    if (threadIdx.x >= shift) {
+      T val = data[offset];
```
Comment (malfet) [inline aten/src/ATen/native/cuda/SoftMax.cu:481]:
```suggestion
      T val = data[threadIdx.x];
```

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 4]
```diff
@@ -440,6 +459,57 @@ ilpReduce(index_t shift,
...
+          const T* data,
+          index_t size,
+          AccumT &threadMax,
+          AccumT &threadExp)
+{
+  MaxFloat<T, AccumT> redFuncMax;
+  Add<AccumT> redFuncAdd;
+
+  // Prologue
+  index_t offset = threadIdx.x;
```
Comment (malfet) [inline aten/src/ATen/native/cuda/SoftMax.cu:476]: Not really needed here
```suggestion
```

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 5]
```diff
@@ -440,6 +459,57 @@ ilpReduce(index_t shift,
...
+    size -= blockDim.x;
+    data += blockDim.x;
+  }
+
+  using LoadT = at::native::memory::aligned_vector<T, ILP>;
+  const LoadT* const data_vec_ptr = reinterpret_cast<const LoadT*>(data);
+
+  // Used vectorized load instructions for the center of the buffer
+  index_t last = size % (ILP * blockDim.x);
+  for (; offset * ILP < (size - last); offset += blockDim.x) {
```
Comment (malfet) [inline aten/src/ATen/native/cuda/SoftMax.cu:494]: Let's keep for for loop in a tight form
```suggestion
  for (index_t offset = threadIdx.x; offset * ILP < (size - last); offset += blockDim.x) {
```

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 6]
```diff
@@ -440,6 +459,57 @@ ilpReduce(index_t shift,
...
+    #pragma unroll
+    for (int j = 0; j < ILP; ++j) {
+      threadMax = redFuncMax(threadMax, v.val[j]);
+      threadExp = redFuncAdd(threadExp, std::exp(v.val[j]));
+    }
+  }
+
+  // Epilogue
+  offset = size - last + threadIdx.x;
+  for (; offset < size; offset += blockDim.x) {
```
Comment (malfet) [inline aten/src/ATen/native/cuda/SoftMax.cu:506]:
```suggestion
  for (index_t offset = size - last + threadIdx.x; offset < size; offset += blockDim.x) {
```

**Review by malfet — COMMENTED** [review]
If ILP==1 first element will be counted twice
if ILP is greater than 1, shouldn't offset in the main loop be denied to something like `(threadIdx.x + ILP -1 ) / ILP`?

**Review by malfet — APPROVED** [review]

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 7]
```diff
@@ -663,6 +698,78 @@ cunn_SoftMaxForward(outscalar_t *output, const scalar_t *input, int classes)
   }
 }
 
+template <int ILP, typename scalar_t, typename accscalar_t, typename outscalar_t, template <typename, typename, typename> class Epilogue>
+__global__ void
+cunn_SoftMaxForwardSmem(outscalar_t *output, const scalar_t *input, int classes)
```
Comment (malfet) [inline aten/src/ATen/native/cuda/SoftMax.cu:703]: I'm not sure why we use `can_use_32bit_indexing` for some of the kernels, but not for others?
```suggestion
template <int ILP, typename scalar_t, typename accscalar_t, typename outscalar_t, template <typename, typename, typename> class Epilogue, typename index_t = int>
__global__ void
cunn_SoftMaxForwardSmem(outscalar_t *output, const scalar_t *input, index_t classes)
```

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 8]
```diff
@@ -116,6 +117,10 @@ inline dim3 SpatialSoftMax_getGridSize(
 
 const int max_threads = 1024;
 
+// The Nvidia Kepler GPUs have 48KB of shmem per SM. Future generations
+// have more. We use the minimum here to preserve some backwards compatibility.
+const int max_smem_per_block = 48 * 1024;
```
Comment (malfet) [inline aten/src/ATen/native/cuda/SoftMax.cu:122]: Wouldn't it be better to query? Also, not sure if this statement is true for ROCm GPUs
cc: @jeffdaily

Comment (valentinandrei) [reply aten/src/ATen/native/cuda/SoftMax.cu:122]: Yep, let me see how we do it for other kernels and if there's no precedent I'll do a query and use a thread local storage to cache.

Comment (valentinandrei) [reply aten/src/ATen/native/cuda/SoftMax.cu]: Noticed that many other kernels use `at::cuda::getCurrentDeviceProperties()->sharedMemPerBlock` so I just added the same query.
